##COLLEGE OF ENGINEERING, TRIVANDRUM
##DEPARTMENT OF COMPUTER SCIENCE & ENGINEERING
##S5 CSE – FIRST ASSIGNMENT, 19th FEBRUARY 2022
##CST 395:NEURAL NETWORKS AND DEEP LEARNING
Aishwarya A J<br>
TVE19CS006<br>
S5 CS<br>
Access the Github Repo From : [This Github Repo](https://github.com/aish2002/NNDL-Assignment-1-TVE19CS006)



**PROBLEM STATEMENT**

Design a simple Neural Network on MNIST Handwritten Digit Dataset. You can
decide the number of layers and number of neurons in your network. Use cross
entropy loss function with adam optimizer for your model. The performance
measurement can be done in terms of accuracy. Compare the performances of
at least two different models created by you by varying the number of layers
and number of neurons in hidden layers. Also write an analysis of your two
models in terms of number of parameters and epochs.

In [20]:
from numpy import mean
from numpy import std
from matplotlib import pyplot as plt
from sklearn.model_selection import KFold
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D
from tensorflow.keras.layers import MaxPooling2D
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import Flatten
from tensorflow.keras.optimizers import Adam

In [21]:
# load dataset
(trainX, trainy), (testX, testy) = mnist.load_data()
# summarize loaded dataset
print('Train: X=%s, y=%s' % (trainX.shape, trainy.shape))
print('Test: X=%s, y=%s' % (testX.shape, testy.shape))

Train: X=(60000, 28, 28), y=(60000,)
Test: X=(10000, 28, 28), y=(10000,)


In [22]:
(trainX, trainY), (testX, testY) = mnist.load_data()

# reshape dataset to have a single channel
trainX = trainX.reshape((trainX.shape[0], 28, 28, 1))
testX = testX.reshape((testX.shape[0], 28, 28, 1))

In [23]:
from tensorflow.keras.utils import to_categorical

# one hot encode target values
trainY = to_categorical(trainY)
testY = to_categorical(testY)

In [24]:
# convert from integers to floats
train_norm = trainX.astype('float32')
test_norm = testX.astype('float32')
# normalize to range 0-1
trainX = train_norm / 255.0
testX = test_norm / 255.0

First NN Model with 2 Layers

In [30]:
def define_model_1():
	model = Sequential()
	model.add(Flatten(input_shape=(28, 28)))
	model.add(Dense(260, activation='relu'))
	model.add(Dense(10, activation='softmax'))
	# compile model
	opt = Adam(learning_rate=0.01)
	model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
	return model

In [31]:
# evaluate a model using k-fold cross-validation
def evaluate_model_1():
    scores, histories = list(), list()
    # define model
    model = define_model_1()
    print(model.summary())
    # fit model
    history = model.fit(trainX, trainY, epochs=10, batch_size=32, validation_data=(testX, testY), verbose=1)
    # evaluate model
    _, acc = model.evaluate(testX, testY, verbose=1)
    print('> %.3f' % (acc * 100.0))
    # stores scores
    scores.append(acc)
    histories.append(history)
    return scores, histories

In [32]:
scores, histories = evaluate_model_1()

Model: "sequential_6"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_6 (Flatten)         (None, 784)               0         
                                                                 
 dense_14 (Dense)            (None, 260)               204100    
                                                                 
 dense_15 (Dense)            (None, 10)                2610      
                                                                 
Total params: 206,710
Trainable params: 206,710
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/10
1875/1875 [==============================] - 7s 3ms/step - loss: 0.2432 - accuracy: 0.9283 - val_loss: 0.2230 - val_accuracy: 0.9374
Epoch 2/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1647 - accuracy: 0.9546 - val_loss: 0.1617 - val_accuracy: 0.9559
Epoch 3/10
1875/187

Second NN Model with 4 Layers

In [43]:
def define_model_2():
    model = Sequential()
    model.add(Flatten(input_shape=(28, 28)))
    model.add(Dense(260, activation='relu'))
    model.add(Dense(140, activation='relu'))
    model.add(Dense(78, activation='relu'))
    model.add(Dense(10, activation='softmax'))
    # compile model
    opt = Adam(learning_rate=0.01)
    model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
    return model

In [44]:
# evaluate a model using k-fold cross-validation
def evaluate_model_2():
    scores, histories = list(), list()
    # define model
    model = define_model_2()
    print(model.summary())
    # fit model
    history = model.fit(trainX, trainY, epochs=10, batch_size=32, validation_data=(testX, testY), verbose=1)
    # evaluate model
    _, acc = model.evaluate(testX, testY, verbose=1)
    print('> %.3f' % (acc * 100.0))
    # stores scores
    scores.append(acc)
    histories.append(history)
    return scores, histories

Execute the Second Model Training

In [45]:
scores, histories = evaluate_model_2()

Model: "sequential_10"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_10 (Flatten)        (None, 784)               0         
                                                                 
 dense_28 (Dense)            (None, 260)               204100    
                                                                 
 dense_29 (Dense)            (None, 140)               36540     
                                                                 
 dense_30 (Dense)            (None, 78)                10998     
                                                                 
 dense_31 (Dense)            (None, 10)                790       
                                                                 
Total params: 252,428
Trainable params: 252,428
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/10
1875/1875 [====================

# Analysis Report

2 Deep Learning Models using Neural Network architecture on the MNIST dataset were developed. First one has 2 NN layers and the second model uses 4 NN layers both with softmax layer activation at the top and relu activation function on other layers. <br> <br>

For the first model,

Number of Layers = 4

Number of Trainable Parameters : 206,710

Validation Accuracy = 96.72% 

Validation Loss = 0.2347 <br> <br>


For the second model, 

Number of Layers = 4

Validation Accuracy = 96.92% 

Validation Loss of = 0.1816

Number of Trainable Parameters :252,428

Clearly Second Model with better validation accuracy is having better performance.
<br><br>

**Submitted by,**

**Aishwarya A J**

**S5 CSE**

**TVE19CS006**
